### ISSUE TITLE
PHP error thrown in Customizer when using `object-fit="cover"` attribute and styles

### ISSUE THREAD

**Issue by laurelfulford** [Automattic/amp-wp#6748]
### Bug Description

When you have an image styled with `object-fit: cover` (including the `object-fit="cover"` attribute on the `img` tag), when you visit the Customizer, it throws a PHP error:

```
Failed to prepare AMP page

A PHP error occurred while trying to prepare the AMP response. This may not be caused by the AMP plugin but by some other active plugin or the current theme. You will need to review the error details to determine the source of the error.

If you get stuck, you may want to share any details in a new topic on the plugin's support forum.

Maximum allowed CSS byte count exceeded for inline style 'object-fit:cover': <img class="i-amphtml-fill-content i-amphtml-replaced-content" decoding="async" src="https://via.placeholder.com/600x600"></img> (0) [AmpProject\Exception\MaxCssByteCountExceeded]
/srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-content/plugins/amp/vendor/ampproject/amp-toolbox/src/Exception/MaxCssByteCountExceeded.php:28

#0 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-content/plugins/amp/vendor/ampproject/amp-toolbox/src/Dom/Element.php(82): AmpProject\Exception\MaxCssByteCountExceeded::forInlineStyle(Object(AmpProject\Dom\Element), 'object-fit:cove...')
#1 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-content/plugins/amp/vendor/ampproject/amp-toolbox/src/Dom/Element.php(61): AmpProject\Dom\Element->setAttribute('style', 'object-fit:cove...')
#2 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-content/plugins/amp/vendor/ampproject/amp-toolbox/src/Optimizer/Transformer/OptimizeHeroImages.php(586): AmpProject\Dom\Element->addInlineStyle('object-fit:cove...')
#3 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-content/plugins/amp/vendor/ampproject/amp-toolbox/src/Optimizer/Transformer/OptimizeHeroImages.php(186): AmpProject\Optimizer\Transformer\OptimizeHeroImages->generateImg(Object(AmpProject\Optimizer\HeroImage), Object(AmpProject\Dom\Document))
#4 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-content/plugins/amp/vendor/ampproject/amp-toolbox/src/Optimizer/TransformationEngine.php(65): AmpProject\Optimizer\Transformer\OptimizeHeroImages->transform(Object(AmpProject\Dom\Document), Object(AmpProject\Optimizer\ErrorCollection))
#5 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-content/plugins/amp/src/Optimizer/OptimizerService.php(46): AmpProject\Optimizer\TransformationEngine->optimizeDom(Object(AmpProject\Dom\Document), Object(AmpProject\Optimizer\ErrorCollection))
#6 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-content/plugins/amp/includes/class-amp-theme-support.php(2056): AmpProject\AmpWP\Optimizer\OptimizerService->optimizeDom(Object(AmpProject\Dom\Document), Object(AmpProject\Optimizer\ErrorCollection))
#7 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-content/plugins/amp/includes/class-amp-theme-support.php(1678): AMP_Theme_Support::prepare_response('<!doctype html>...')
#8 [internal function]: AMP_Theme_Support::finish_output_buffering('<!doctype html>...', 9)
#9 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-includes/functions.php(5107): ob_end_flush()
#10 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-includes/class-wp-hook.php(303): wp_ob_end_flush_all('')
#11 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-includes/class-wp-hook.php(327): WP_Hook->apply_filters('', Array)
#12 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-includes/plugin.php(470): WP_Hook->do_action(Array)
#13 /srv/users/usercd1e91d4/apps/usercd1e91d4/public/wp-includes/load.php(1097): do_action('shutdown')
#14 [internal function]: shutdown_action_hook()
#15 {main}
```

It looks like this started happening as of version 2.1.3 of the plugin; it works as expected in versions 2.1.2 and older. 

I originally ran into this with the [Newspack Blocks](https://github.com/automattic/newspack-blocks) plugin -- the Homepage Posts block lets you set the [featured image behind](https://newspack.pub/support/blocks/homepage-posts/#media-location) the block content. But I can also recreate it on a test site using Twenty Twenty-One, by adding the following HTML to a Custom HTML block, and then opening the Customizer from that post. To recreate, you can keep the `src` the same, or change it to one in your Media Library:

```
<div style="height: 500px; width: 500px; overflow: hidden">
<img style="object-fit: cover; display: block; width: 100%; height: 100%" src="https://via.placeholder.com/600x600" object-fit="cover">
</div>
```

### Expected Behaviour

I would expect this attribute to not throw errors. 

In experimenting with the markup further, it looks like the attribute may no longer be required to make images display cover when AMP is enabled, but I haven't been able to find confirmation (my search skills are probably to blame here!). 

### Screenshots

![image](https://user-images.githubusercontent.com/177561/143622956-5a58bc2b-e1dd-401a-acca-ec2ee51c5aa5.png)

A Gist of the front-end of the page can be found [here](https://gist.github.com/laurelfulford/e9b0c3615cab53cdb120606045329623). 

### PHP Version

7.4.26

### Plugin Version

2.1.3 and 2.1.4

### AMP plugin template mode

Standard

### WordPress Version

5.8.2

**Comment by westonruter** [comment]
Thanks for reporting! This is something I ran into as well in another context: https://github.com/ampproject/amp-toolbox-php/pull/319.

The issue is that when AMP Optimizer was performing server-side rendering (SSR), it wasn't accounting for the fact that Optimized AMP pages have a higher CSS byte limit due to the additional CSS which is generated (such as for `object-fit`). 

Would you please re-test your issue in the latest 2.2-alpha build available from our the [Development Builds](https://github.com/ampproject/amp-wp/wiki/Development-Builds) wiki page?

**Comment by westonruter** [comment]
Also, I presume this is is happening in an “AMP Plus” context where there is more than 75KB of CSS being added to the page.

**Comment by westonruter** [comment]
Actually, I see that it was previously increasing the max byte count for SSR'ed pages:

![image](https://user-images.githubusercontent.com/134745/143991844-34a1ea22-857d-4450-aa3e-2693d79ee0a9.png)

So this may still be an issue.

**Comment by westonruter** [comment]
I'll investigate further. However…

> In experimenting with the markup further, it looks like the attribute may no longer be required to make images display cover when AMP is enabled, but I haven't been able to find confirmation (my search skills are probably to blame here!).

As a workaround, you can do the styling with CSS instead. You can set the `object-fit` CSS style (i.e. via stylesheet) on an `amp-img` and it will now get inherited by the `img`. So you don't need to set the `object-fit` HTML attribute on an `amp-img` anymore. This landed in AMP v2.1 via https://github.com/ampproject/amp-wp/pull/5955. I see your page is running AMP v2.0.11, so upgrade to the latest to ensure you have this available.

**Comment by laurelfulford** [comment]
Thanks for all this extra information, @westonruter!

> Would you please re-test your issue in the latest 2.2-alpha build available from our the Development Builds wiki page?

I'm not sure if this information was still needed, but I can confirm that the 2.2-alpha version of the plugin did fix this issue for me! 

> I see your page is running AMP v2.0.11, so upgrade to the latest to ensure you have this available.

I'm afraid I grabbed that gist when I was running the older version of the plugin, trying to figure out when `object-fit` stopped being required. Thank you for confirming it's not with the newer versions of the plugin. 

As this looks like it'll be resolved in an upcoming version of the plugin, just let me know if you'd like me to close the issue, or leave it open for further investigation -- thank you!

**Comment by westonruter** [comment]
I can confirm this issue is fixed in 2.2. If I patch the plugin as follows:

```diff
diff --git a/includes/class-amp-theme-support.php b/includes/class-amp-theme-support.php
index 135c84c24..094062e3a 100644
--- a/includes/class-amp-theme-support.php
+++ b/includes/class-amp-theme-support.php
@@ -2126,11 +2126,7 @@ public static function prepare_response( $response, $args = [] ) {
 
 			$errors = new Optimizer\ErrorCollection();
 
-			$args['skip_css_max_byte_count_enforcement'] = (
-				$has_validation_exemptions
-				||
-				DevMode::isActiveForDocument( $dom )
-			);
+			$args['skip_css_max_byte_count_enforcement'] = false;
 			self::get_optimizer( $args )->optimizeDom( $dom, $errors );
 
 			if ( count( $errors ) > 0 ) {
```

I was able to reproduce this issue by activating the following plugin:

```php
<?php
/**
 * Plugin Name: Try Object-Fit Error
 */

add_action( 'wp_footer', function () {
	echo '<img src="https://amp-wp.org/wp-content/uploads/2020/08/Asset-25@2x-1.png" data-hero width="200" height="200" object-fit="cover">';

	$pad_count = isset( $_GET['css_padding'] ) ? (int) $_GET['css_padding'] : 0;

	printf( '<style>body:after { display:none; content: "%s";  }</style>', str_repeat( 'a', $pad_count ) );
} );
```

With a site in Standard mode, I tried going to the URL `/?css_padding=100000`. I would then get the error thrown.

This error is prevented by opting-out of enforcing the CSS limit when a page is in dev mode (which is the case in the Customizer) or it has validation exemptions (as in the case of excessive CSS in Standard mode). See bdc673ab7ce933baff5df800b1b87c2ab7c0b74d and d615b570bc097d05e3a61cd4a79db560d05162c9.

Thanks for reporting and confirming the issue is fixed in 2.2-alpha! We are working to release 2.2 stable this week.
